mpc/mpc-aio/src/protocol/garble/exec/dual.rs

```diff
@@ -12,287 +12,570 @@ use crate::protocol::{
+        let msg = expect_msg_or_err!(
+            self.channel.next().await,
+            GarbleMessage::GarbledCircuit,
+            GCError::Unexpected
+        )?;
+
+        let gc_ev =
+            GarbledCircuit::<gc_state::Partial>::from_unchecked(self.circ.clone(), msg.into())?;
+
+        // Evaluate garbled circuit
+        let evaluated_gc = self.backend.evaluate(gc_ev, self.state.ev_labels).await?;
+
+        let leader = leader.from_evaluated_circuit(evaluated_gc)?;
+
+        // Commit to output labels
+        let (commit, leader) = leader.commit();
+
+        // Send commitment
+        self.channel
+            .send(GarbleMessage::HashCommitment(commit.into()))
+            .await?;
+
+        // Expect output labels digest from follower
+        let msg = expect_msg_or_err!(
+            self.channel.next().await,
+            GarbleMessage::OutputLabelsDigest,
+            GCError::Unexpected
+        )?;
+
+        // Perform equality check
+        let follower_check = msg.into();
+        let leader = leader.check(follower_check)?;
+
+        // If equality check passes, reveal output digest
+        let (opening, gc_evaluated) = leader.reveal();
+
+        self.channel
+            .send(GarbleMessage::CommitmentOpening(opening.into()))
+            .await?;
+
+        Ok(gc_evaluated.summarize())
+    }
+}
+
+pub struct DualExFollower<S, B, LS, LR>
+where
+    S: State,
     B: Generator + Evaluator,
-    S: ObliviousSend<FullInputLabels>,
-    R: ObliviousReceive<InputValue, ActiveInputLabels>,
+    LS: ObliviousSend<FullInputLabels>,
+    LR: ObliviousReceive<InputValue, ActiveInputLabels>,
 {
+    state: S,
+    circ: Arc<Circuit>,
     channel: GarbleChannel,
     backend: B,
-    label_sender: S,
-    label_receiver: R,
+    label_sender: Option<LS>,
+    label_receiver: Option<LR>,
 }
 
-impl<B, S, R> DualExFollower<B, S, R>
+impl<B, LS, LR> DualExFollower<Initialized, B, LS, LR>
 where
     B: Generator + Evaluator + Send,
-    S: ObliviousSend<FullInputLabels> + Send,
-    R: ObliviousReceive<InputValue, ActiveInputLabels> + Send,
+    LS: ObliviousSend<FullInputLabels> + Send,
+    LR: ObliviousReceive<InputValue, ActiveInputLabels> + Send,
 {
-    pub fn new(channel: GarbleChannel, backend: B, label_sender: S, label_receiver: R) -> Self {
-        Self {
+    /// Create a new DualExFollower
+    pub fn new(
+        circ: Arc<Circuit>,
+        channel: GarbleChannel,
+        backend: B,
+        label_sender: Option<LS>,
+        label_receiver: Option<LR>,
+    ) -> DualExFollower<Initialized, B, LS, LR> {
+        DualExFollower {
+            state: Initialized,
+            circ,
             channel,
             backend,
             label_sender,
             label_receiver,
         }
     }
+
+    /// Exchange input labels
+    ///
+    /// * `gen_labels` - Labels to garble the follower's circuit
+    /// * `gen_inputs` - Inputs to be sent directly to the leader
```

Comment:
/// * `gen_inputs` - Inputs for which the labels are to be sent directly to the leader
    /// * `ot_send_inputs` - Inputs for which the labels are to be sent via OT
    /// * `ot_receive_inputs` - Inputs for which the labels are to be received via OT
    /// * `cached_labels` - Cached input labels for the leader's circuit. We can cache both our
    ///                     active input labels and also the leaders's active input labels.